File: java/src/main/java/com/example/employeeschedulemanagementsys/gui/MainFrame.java
package com.example.employeeschedulemanagementsys.gui;

import com.example.employeeschedulemanagementsys.model.Employee;
import com.example.employeeschedulemanagementsys.service.SchedulingService;

import javax.swing.*;
import javax.swing.border.EmptyBorder;
import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class MainFrame extends JFrame {

    private final List<Employee> employees = new ArrayList<>();
    private final SchedulingService service = new SchedulingService();

    private final DefaultListModel<String> listModel = new DefaultListModel<>();
    private final JList<String> empList = new JList<>(listModel);
    private final JTextField nameField = new JTextField(14);
    private final JTextArea outputArea = new JTextArea();

    public MainFrame() {
        super("Employee Shift Scheduler");
        setDefaultCloseOperation(EXIT_ON_CLOSE);
        setSize(1100, 560);
        setMinimumSize(new Dimension(700, 420));
        setLocationRelativeTo(null);

        JSplitPane split = new JSplitPane(
                JSplitPane.HORIZONTAL_SPLIT,
                buildInputPanel(),
                buildOutputPanel());
        split.setDividerLocation(370);
        split.setResizeWeight(0.0);
        split.setBorder(new EmptyBorder(8, 8, 8, 8));
        setContentPane(split);
    }

    private JPanel buildInputPanel() {

        // FlowLayout keeps each component at its natural preferred size.
        // The previous BorderLayout.EAST was squishing "Add" to a tiny sliver.
        JPanel nameRow = new JPanel(new FlowLayout(FlowLayout.LEFT, 4, 0));
        nameRow.setBorder(new EmptyBorder(0, 0, 6, 0));
        nameRow.add(new JLabel("Name:"));
        nameField.setPreferredSize(new Dimension(148, 26));
        nameRow.add(nameField);
        JButton addBtn = btn("Add", new Color(34, 139, 34));
        addBtn.setPreferredSize(new Dimension(56, 26));
        addBtn.addActionListener(e -> addEmployee());
        nameField.addActionListener(e -> addEmployee());
        nameRow.add(addBtn);

        empList.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
        empList.setFont(new Font(Font.SANS_SERIF, Font.PLAIN, 13));
        empList.setFixedCellHeight(24);
        empList.addMouseListener(new java.awt.event.MouseAdapter() {
            @Override
            public void mouseClicked(java.awt.event.MouseEvent e) {
                if (e.getClickCount() == 2) openPreferences();
            }
        });

        JScrollPane listScroll = new JScrollPane(empList);
        listScroll.setBorder(BorderFactory.createTitledBorder("Employees  (double-click to edit)"));
        listScroll.setPreferredSize(new Dimension(250, 300));

        JButton prefsBtn = btn("Set Preferences", new Color(25, 100, 200));
        JButton removeBtn = btn("Remove Selected", new Color(180, 40, 40));
        JButton genBtn = btn("Generate Schedule", new Color(20, 20, 20));

        prefsBtn.addActionListener(e -> openPreferences());
        removeBtn.addActionListener(e -> removeEmployee());
        genBtn.addActionListener(e -> generateSchedule());

        JPanel btnStack = new JPanel(new GridLayout(3, 1, 0, 5));
        btnStack.setBorder(new EmptyBorder(6, 0, 0, 0));
        btnStack.add(prefsBtn);
        btnStack.add(removeBtn);
        btnStack.add(genBtn);

        JPanel panel = new JPanel(new BorderLayout(0, 4));
        panel.setBorder(new EmptyBorder(4, 4, 4, 4));
        panel.add(nameRow, BorderLayout.NORTH);
        panel.add(listScroll, BorderLayout.CENTER);
        panel.add(btnStack, BorderLayout.SOUTH);
        return panel;
    }

    private JPanel buildOutputPanel() {
        outputArea.setEditable(false);
        outputArea.setFont(new Font(Font.MONOSPACED, Font.PLAIN, 12));
        outputArea.setMargin(new Insets(6, 8, 6, 8));
        outputArea.setText(
                "How to use:\n" +
                        "  1. Type a name and click Add (or press Enter).\n" +
                        "  2. The new employee is auto-selected in the list.\n" +
                        "  3. Click \"Set Preferences\" (or double-click the name).\n" +
                        "  4. Repeat for each employee (need at least 2).\n" +
                        "  5. Click \"Generate Schedule\".\n");

        JScrollPane scroll = new JScrollPane(outputArea);
        scroll.setBorder(BorderFactory.createTitledBorder("Schedule Output"));

        JPanel panel = new JPanel(new BorderLayout());
        panel.add(scroll, BorderLayout.CENTER);
        return panel;
    }

    private JButton btn(String text, Color bg) {
        JButton b = new JButton(text);
        b.setBackground(bg);
        b.setForeground(Color.WHITE);
        b.setOpaque(true);
        b.setBorderPainted(false);
        b.setFocusPainted(false);
        b.setFont(new Font(Font.SANS_SERIF, Font.BOLD, 12));
        return b;
    }

    private void addEmployee() {
        String name = nameField.getText().trim();
        if (name.isEmpty()) {
            warn("Enter a name first.");
            return;
        }
        for (Employee e : employees) {
            if (e.getName().equalsIgnoreCase(name)) {
                warn("\"" + name + "\" is already in the list.");
                return;
            }
        }
        employees.add(new Employee(name));
        listModel.addElement(name);
        nameField.setText("");
        empList.setSelectedIndex(listModel.getSize() - 1);
    }

    private void removeEmployee() {
        int idx = empList.getSelectedIndex();
        if (idx < 0) {
            warn("Select an employee from the list first.");
            return;
        }
        employees.remove(idx);
        listModel.remove(idx);
    }

    private void openPreferences() {
        int idx = empList.getSelectedIndex();
        if (idx < 0) {
            warn("Select an employee from the list first.");
            return;
        }
        new PreferenceDialog(this, employees.get(idx)).setVisible(true);
    }

    private void generateSchedule() {
        if (employees.size() < 2) {
            warn("Add at least 2 employees before generating.");
            return;
        }
        service.buildSchedule(employees);
        outputArea.setText(service.buildReport(employees));
        outputArea.setCaretPosition(0);
    }

    private void warn(String msg) {
        JOptionPane.showMessageDialog(this, msg, "Notice", JOptionPane.WARNING_MESSAGE);
    }
}

File: jsapp/gui/MainFrame.js
import Employee         from "../model/Employee.js";
import SchedulingService from "../service/SchedulingService.js";
import PreferenceDialog  from "./PreferenceDialog.js";

export default class MainFrame {

    constructor() {
        this.employees     = [];       // mirrors: List<Employee> employees
        this.service       = new SchedulingService();
        this.selectedIndex = -1;       // mirrors: empList.getSelectedIndex()

        this._buildDOM();
        this._bindEvents();
    }

    _buildDOM() {
        // Root container
        const app = document.getElementById("app");

        //  Left panel
        const left = document.createElement("div");
        left.id = "left-panel";

        // Name row
        const nameRow = document.createElement("div");
        nameRow.id = "name-row";

        const nameLabel = document.createElement("label");
        nameLabel.textContent = "Name:";
        nameLabel.htmlFor     = "nameInput";

        this.nameInput = document.createElement("input");
        this.nameInput.id          = "nameInput";
        this.nameInput.type        = "text";
        this.nameInput.placeholder = "type name…";
        this.nameInput.autocomplete = "off";

        this.addBtn = this._btn("Add", "btn-green");

        nameRow.appendChild(nameLabel);
        nameRow.appendChild(this.nameInput);
        nameRow.appendChild(this.addBtn);

        // Employee list — mirrors JScrollPane + JList
        const listBox = document.createElement("div");
        listBox.id = "list-box";

        const listLegend = document.createElement("span");
        listLegend.className   = "box-legend";
        listLegend.textContent = "Employees  (double-click to edit)";

        this.empListEl = document.createElement("div");
        this.empListEl.id = "empList";

        listBox.appendChild(listLegend);
        listBox.appendChild(this.empListEl);

        // Action buttons — mirrors 3 JButtons in GridLayout
        const btnStack = document.createElement("div");
        btnStack.id = "btn-stack";

        this.prefsBtn  = this._btn("Set Preferences",  "btn-blue");
        this.removeBtn = this._btn("Remove Selected",   "btn-red");
        this.genBtn    = this._btn("Generate Schedule", "btn-black");

        btnStack.appendChild(this.prefsBtn);
        btnStack.appendChild(this.removeBtn);
        btnStack.appendChild(this.genBtn);

        left.appendChild(nameRow);
        left.appendChild(listBox);
        left.appendChild(btnStack);

        // ── Right panel ──────────────────────────────────────────────────
        const right = document.createElement("div");
        right.id = "right-panel";

        const outputLegend = document.createElement("span");
        outputLegend.className   = "box-legend";
        outputLegend.textContent = "Schedule Output";

        // mirrors: JTextArea outputArea (read-only, monospaced)
        this.outputArea = document.createElement("div");
        this.outputArea.id = "output";
        this.outputArea.textContent = [
            "How to use:",
            "  1. Type a name and click Add (or press Enter).",
            "  2. The new employee is auto-selected in the list.",
            '  3. Click "Set Preferences" (or double-click the name).',
            "  4. Repeat for each employee (need at least 2).",
            '  5. Click "Generate Schedule".'
        ].join("\n");

        right.appendChild(outputLegend);
        right.appendChild(this.outputArea);

        app.appendChild(left);
        app.appendChild(right);

        // ── Alert box — mirrors JOptionPane.showMessageDialog ────────────
        this.alertOverlay = document.createElement("div");
        this.alertOverlay.id = "alert-overlay";

        const alertBox = document.createElement("div");
        alertBox.id = "alert-box";

        this.alertMsg = document.createElement("div");
        this.alertMsg.id = "alert-msg";

        const alertOk = document.createElement("button");
        alertOk.id          = "alert-ok";
        alertOk.textContent = "OK";
        alertOk.addEventListener("click", () =>
            this.alertOverlay.classList.remove("open")
        );

        alertBox.appendChild(this.alertMsg);
        alertBox.appendChild(alertOk);
        this.alertOverlay.appendChild(alertBox);
        document.body.appendChild(this.alertOverlay);
    }

    // ── Bind events ────────────────────────────────────────────────────

    /**
     * Wires up all button listeners and keyboard shortcuts.
     * mirrors: addActionListener calls in MainFrame.java constructor
     */
    _bindEvents() {
        // Add button + Enter key
        // mirrors: addBtn.addActionListener(e -> addEmployee())
        //          nameField.addActionListener(e -> addEmployee())
        this.addBtn  .addEventListener("click",   () => this.addEmployee());
        this.nameInput.addEventListener("keydown", e => {
            if (e.key === "Enter") this.addEmployee();
        });

        // mirrors: prefsBtn.addActionListener(e -> openPreferences())
        this.prefsBtn .addEventListener("click", () => this.openPreferences());

        // mirrors: removeBtn.addActionListener(e -> removeEmployee())
        this.removeBtn.addEventListener("click", () => this.removeEmployee());

        // mirrors: genBtn.addActionListener(e -> generateSchedule())
        this.genBtn   .addEventListener("click", () => this.generateSchedule());

        // PreferenceDialog delegates warn() via a custom event
        document.addEventListener("app:warn", e => this._warn(e.detail));
    }

    // ── Actions ────────────────────────────────────────────────────────

    /**
     * mirrors: addEmployee()
     */
    addEmployee() {
        const name = this.nameInput.value.trim();
        if (!name) { this._warn("Enter a name first."); return; }

        // Duplicate check — mirrors equalsIgnoreCase loop
        for (const e of this.employees) {
            if (e.name.toLowerCase() === name.toLowerCase()) {
                this._warn(`"${name}" is already in the list.`);
                return;
            }
        }

        this.employees.push(new Employee(name));
        this.nameInput.value = "";

        // Auto-select — mirrors empList.setSelectedIndex(listModel.getSize() - 1)
        this.selectedIndex = this.employees.length - 1;
        this._renderList();
    }

    /**
     * mirrors: removeEmployee()
     */
    removeEmployee() {
        if (this.selectedIndex < 0) {
            this._warn("Select an employee from the list first.");
            return;
        }
        this.employees.splice(this.selectedIndex, 1);
        this.selectedIndex = Math.min(this.selectedIndex, this.employees.length - 1);
        this._renderList();
    }

    /**
     * mirrors: openPreferences() → new PreferenceDialog(this, employees.get(idx)).setVisible(true)
     */
    openPreferences() {
        if (this.selectedIndex < 0) {
            this._warn("Select an employee from the list first.");
            return;
        }
        const emp    = this.employees[this.selectedIndex];
        const dialog = new PreferenceDialog(emp, () => {});
        dialog.show();
    }

    /**
     * mirrors: generateSchedule()
     */
    generateSchedule() {
        if (this.employees.length < 2) {
            this._warn("Add at least 2 employees before generating.");
            return;
        }
        this.service.buildSchedule(this.employees);
        this.outputArea.textContent = this.service.buildReport(this.employees);
    }

    // ── Rendering ──────────────────────────────────────────────────────

    /**
     * mirrors: JList rendering via DefaultListModel
     * Rebuilds the employee list DOM from this.employees[].
     */
    _renderList() {
        this.empListEl.innerHTML = "";

        for (let i = 0; i < this.employees.length; i++) {
            const item = document.createElement("div");
            item.className   = "emp-item" + (i === this.selectedIndex ? " selected" : "");
            item.textContent = this.employees[i].name;

            // Single click → select
            // mirrors: empList selection listener
            item.addEventListener("click", () => {
                this.selectedIndex = i;
                this._renderList();
            });

            // Double click → open preferences
            // mirrors: MouseAdapter mouseClicked with getClickCount() == 2
            item.addEventListener("dblclick", () => {
                this.selectedIndex = i;
                this._renderList();
                this.openPreferences();
            });

            this.empListEl.appendChild(item);
        }
    }

    // ── Helpers ────────────────────────────────────────────────────────

    /**
     * mirrors: warn(String msg) → JOptionPane.showMessageDialog(...)
     */
    _warn(msg) {
        this.alertMsg.textContent = msg;
        this.alertOverlay.classList.add("open");
    }

    /**
     * mirrors: btn(String text, Color bg) — styled button factory
     */
    _btn(text, colorClass) {
        const b = document.createElement("button");
        b.className   = `btn ${colorClass}`;
        b.textContent = text;
        return b;
    }
}